File: code-recipes/python/src/recipes/vendor_retail_procurement/transaction/get_transaction_recipe.py
"""
Vendor Transaction Status API: Get Transaction
================================================

Checks the status of asynchronous POST transactions using the
Vendor Transaction Status API.

Use cases:
- Check if a submitAcknowledgement was processed successfully
- Check if a submitShipmentConfirmations was processed
- Check if a submitInvoices was processed
- Monitor async operation completion

API operation: getTransaction
"""

import logging
import time
from typing import Any, Dict, Optional

from spapi import SPAPIClient, SPAPIConfig
from spapi.api.vendor_transaction_status_v1.vendor_transaction_api import VendorTransactionApi

from src import config
from src.util.recipe import Recipe

logger = logging.getLogger(__name__)


class GetTransactionRecipe(Recipe):

    def __init__(
        self,
        sp_config: Optional[SPAPIConfig] = None,
        vendor_transaction_api: Optional[VendorTransactionApi] = None,
    ) -> None:
        super().__init__(config=sp_config)
        self._vendor_transaction_api = vendor_transaction_api

    @property
    def vendor_transaction_api(self) -> VendorTransactionApi:
        if self._vendor_transaction_api is None:
            client = SPAPIClient(
                self.config,
                oauth_endpoint=f"{config.backend_url}/auth/o2/token",
                endpoint=config.backend_url,
            )
            self._vendor_transaction_api = VendorTransactionApi(client.api_client)
            logger.info("VendorTransactionApi client initialized")
        return self._vendor_transaction_api

    def get_transaction_status(self, transaction_id: str) -> Optional[Dict[str, Any]]:
        """Get the status of a transaction by its ID."""
        logger.info("Fetching transaction status: %s", transaction_id)
        response = self.vendor_transaction_api.get_transaction(
            transaction_id=transaction_id,
        )
        if hasattr(response, "to_dict"):
            response = response.to_dict()
        payload = response.get("payload") or {}
        transaction = payload.get("transaction_status")
        if transaction:
            logger.info("Retrieved status for transaction: %s", transaction_id)
            return transaction
        logger.warning("No status found for transaction: %s", transaction_id)
        return None

    def is_transaction_complete(self, transaction_id: str) -> bool:
        """Check if a transaction has completed (success or failure)."""
        transaction = self.get_transaction_status(transaction_id)
        if not transaction:
            return False
        status = transaction.get("status")
        return status in ("Success", "Failure")

    def is_transaction_successful(self, transaction_id: str) -> bool:
        """Check if a transaction was successful."""
        transaction = self.get_transaction_status(transaction_id)
        if not transaction:
            return False
        return transaction.get("status") == "Success"

    def poll_for_completion(
        self, transaction_id: str, max_attempts: int = 10, delay_seconds: float = 2.0
    ) -> Optional[Dict[str, Any]]:
        """Poll for transaction completion with timeout."""
        for attempt in range(1, max_attempts + 1):
            transaction = self.get_transaction_status(transaction_id)
            if transaction:
                status = transaction.get("status")
                logger.info(
                    "Attempt %d/%d: Transaction %s status = %s",
                    attempt, max_attempts, transaction_id, status,
                )
                if status in ("Success", "Failure"):
                    return transaction
            if attempt < max_attempts:
                time.sleep(delay_seconds)
        logger.warning(
            "Transaction %s did not complete within %d attempts",
            transaction_id, max_attempts,
        )
        return None

    def start(self) -> None:
        transaction_id = "20190904190535-eef8cad8-418e-4ed3-ac72-789e2ee6214a"
        transaction = self.get_transaction_status(transaction_id)
        if transaction:
            self._log_transaction_status(transaction)

    @staticmethod
    def _log_transaction_status(transaction: Dict[str, Any]) -> None:
        logger.info("========================================")
        logger.info("Transaction Status")
        logger.info("========================================")
        logger.info("Transaction ID: %s", transaction.get("transaction_id"))
        logger.info("Status: %s", transaction.get("status"))
        errors = transaction.get("errors") or []
        if errors:
            logger.info("Errors:")
            for error in errors:
                logger.info("  - Code: %s | Message: %s", error.get("code"), error.get("message"))


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    recipe = GetTransactionRecipe()
    recipe.start()

File: code-recipes/java/src/main/java/vendorretailprocurement/transaction/GetTransactionRecipe.java
package vendorretailprocurement.transaction;

import com.amazon.SellingPartnerAPIAA.LWAException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import software.amazon.spapi.ApiException;
import software.amazon.spapi.api.vendor.transactionstatus.v1.VendorTransactionApi;
import software.amazon.spapi.models.vendor.transactionstatus.v1.GetTransactionResponse;
import software.amazon.spapi.models.vendor.transactionstatus.v1.Transaction;
import util.Constants;
import util.Recipe;

/**
 * Vendor Transaction Status API Recipe: Get Transaction
 * ======================================================
 * 
 * This recipe demonstrates how to check the status of asynchronous
 * POST transactions using the Vendor Transaction Status API.
 * 
 * Use cases:
 * - Check if a submitAcknowledgement was processed successfully
 * - Check if a submitShipmentConfirmations was processed
 * - Check if a submitInvoices was processed
 * - Monitor async operation completion
 * 
 * Transaction statuses:
 * - Processing: Transaction is still being processed
 * - Success: Transaction completed successfully
 * - Failure: Transaction failed (check errors for details)
 * 
 * API Operations used:
 * - getTransaction: Returns the status of a specific transaction
 */
public class GetTransactionRecipe extends Recipe {
    private static final Logger logger = LoggerFactory.getLogger(GetTransactionRecipe.class);

    private final VendorTransactionApi vendorTransactionApi = new VendorTransactionApi.Builder()
            .lwaAuthorizationCredentials(lwaCredentials)
            .endpoint(Constants.BACKEND_URL)
            .build();

    @Override
    public void start() {
        // Example: Check status of a transaction
        String transactionId = "20190904190535-eef8cad8-418e-4ed3-ac72-789e2ee6214a";
        
        Transaction transaction = getTransactionStatus(transactionId);
        if (transaction != null) {
            logTransactionStatus(transaction);
        }
    }

    /**
     * Get the status of a transaction by its ID.
     * 
     * @param transactionId The GUID returned from a POST operation
     * @return Transaction status object, or null if not found
     */
    public Transaction getTransactionStatus(String transactionId) {
        try {
            GetTransactionResponse response = vendorTransactionApi.getTransaction(transactionId);
            
            if (response.getPayload() != null && response.getPayload().getTransactionStatus() != null) {
                logger.info("Retrieved status for transaction: {}", transactionId);
                return response.getPayload().getTransactionStatus();
            }
            logger.warn("No status found for transaction: {}", transactionId);
            return null;
        } catch (ApiException | LWAException e) {
            logger.error("Error getting transaction status for {}: {}", transactionId, e.getMessage(), e);
            throw new RuntimeException("Failed to get transaction status", e);
        }
    }

    /**
     * Check if a transaction has completed (success or failure).
     */
    public boolean isTransactionComplete(String transactionId) {
        Transaction transaction = getTransactionStatus(transactionId);
        if (transaction == null) {
            return false;
        }
        
        String status = transaction.getStatus().getValue();
        return "Success".equals(status) || "Failure".equals(status);
    }

    /**
     * Check if a transaction was successful.
     */
    public boolean isTransactionSuccessful(String transactionId) {
        Transaction transaction = getTransactionStatus(transactionId);
        if (transaction == null) {
            return false;
        }
        
        return "Success".equals(transaction.getStatus().getValue());
    }

    /**
     * Poll for transaction completion with timeout.
     * 
     * @param transactionId The transaction to poll
     * @param maxAttempts Maximum number of polling attempts
     * @param delayMs Delay between attempts in milliseconds
     * @return Final transaction status, or null if timeout
     */
    public Transaction pollForCompletion(String transactionId, int maxAttempts, long delayMs) {
        for (int attempt = 1; attempt <= maxAttempts; attempt++) {
            Transaction transaction = getTransactionStatus(transactionId);
            
            if (transaction != null) {
                String status = transaction.getStatus().getValue();
                logger.info("Attempt {}/{}: Transaction {} status = {}", 
                        attempt, maxAttempts, transactionId, status);
                
                if ("Success".equals(status) || "Failure".equals(status)) {
                    return transaction;
                }
            }
            
            if (attempt < maxAttempts) {
                try {
                    Thread.sleep(delayMs);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    break;
                }
            }
        }
        
        logger.warn("Transaction {} did not complete within {} attempts", transactionId, maxAttempts);
        return null;
    }

    private void logTransactionStatus(Transaction transaction) {
        logger.info("========================================");
        logger.info("Transaction Status");
        logger.info("========================================");
        logger.info("Transaction ID: {}", transaction.getTransactionId());
        logger.info("Status: {}", transaction.getStatus());
        
        if (transaction.getErrors() != null && !transaction.getErrors().isEmpty()) {
            logger.info("Errors:");
            for (var error : transaction.getErrors()) {
                logger.info("  - Code: {} | Message: {}", error.getCode(), error.getMessage());
            }
        }
    }
}
